pkg/controllers/elasticsearch/actions/scale.go

```diff
@@ -0,0 +1,212 @@
+package actions
+
+import (
+	"fmt"
+
+	apps "k8s.io/api/apps/v1beta1"
+	core "k8s.io/api/core/v1"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilerror "k8s.io/apimachinery/pkg/util/errors"
+
+	"github.com/jetstack/navigator/pkg/apis/navigator/v1alpha1"
+	"github.com/jetstack/navigator/pkg/controllers"
+	"github.com/jetstack/navigator/pkg/controllers/elasticsearch/util"
+	utilapi "github.com/jetstack/navigator/pkg/util/api"
+)
+
+type Scale struct {
+	// The Elasticsearch cluster resource being operated on
+	Cluster *v1alpha1.ElasticsearchCluster
+	// The node pool being scaled
+	NodePool *v1alpha1.ElasticsearchClusterNodePool
+	// Number of replicas to scale to
+	Replicas int32
+}
+
+var _ controllers.Action = &Scale{}
+
+func (c *Scale) Name() string {
+	return "Scale"
+}
+
+func (c *Scale) Message() string {
+	return fmt.Sprintf("Scaled node pool %q to %d replicas", c.NodePool.Name, c.NodePool.Replicas)
+}
+
+// Execute will scale the s.NodePool statefulset to the desired number of
+// replicas. It will refuse to scale if the cluster is not in a 'state to scale'
+// as defined by s.canScaleNodePool.
+func (s *Scale) Execute(state *controllers.State) error {
+	if s.NodePool == nil || s.Cluster == nil {
+		return fmt.Errorf("internal error: nodepool and cluster must be set")
+	}
+
+	statefulSetName := util.NodePoolResourceName(s.Cluster, s.NodePool)
+	statefulSet, err := state.StatefulSetLister.StatefulSets(s.Cluster.Namespace).Get(statefulSetName)
+	if k8sErrors.IsNotFound(err) {
+		state.Recorder.Eventf(s.Cluster, core.EventTypeWarning, "Err"+s.Name(), "Cannot get statefulset %q for node pool %q: %v", statefulSetName, s.NodePool.Name, err)
+		return nil
```

Comment:
We return nil to prevent an endless loop in the case where the cluster doesn't exist. Reasoning being that when the StatefulSet *is* added to the listers store, it will trigger the informers OnAdd to fire again, thus re-triggering this.

Reply:
> we should have always created a statefulset before we attempt to scale, right?

Race issues could cause the StatefulSet to no longer exist after the initial check for it existing happens. Also the lister may be stale, which could cause a 404.